borg/archiver.py

```diff
@@ -719,6 +719,134 @@ def do_upgrade(self, args):
+                print("Old size:", format_file_size(self.old_size))
+                print("New size:", format_file_size(self.new_size))
+                if self.old_size:
+                    ratio = self.new_size / self.old_size
+                else:
+                    ratio = 0
+                print("Ratio: %d %%" % (ratio * 100))
+
+        def detect_compressor(compression_header):
+            for compressor in COMPRESSOR_LIST:
+                if compressor.detect(compression_header):
+                    return compressor
+            else:
+                raise ValueError('No decompressor for this data found: %r.', compression_header)
+
+        def recompress_chunk(decompressor, compressed_chunk):
+            decompressed_data = decompressor(buffer=decompress_buffer).decompress(compressed_chunk)
+            key.assert_chunk_id(id_, decompressed_data)
+            return key.encrypt(decompressed_data)
+
+        def do_exit_soon(sig_num, stack_frame):
+            nonlocal exit_soon
+            if exit_soon:
+                sys.stderr.write("Received signal, again. I'm not deaf.\n")
+            else:
+                sys.stderr.write("Received signal, will exit cleanly.\n")
+            sys.stderr.flush()
+            exit_soon = True
+
+        signal.signal(signal.SIGTERM, do_exit_soon)
+        signal.signal(signal.SIGINT, do_exit_soon)
+
+        stats = Stats()
+        exit_soon = False
+
+        dry_run = args.dry_run
+        repository = self.open_repository(args, exclusive=True)
+        manifest, key = Manifest.load(repository)
+        segment_pointer = 0
+        if args.segment_pointer:
+            if args.force_recompress:
+                repository.config.remove_option('repository', 'recompress_segment_pointer')
+            else:
+                segment_pointer = repository.config.getint('repository', 'recompress_segment_pointer', fallback=0)
+        compr_args = dict(buffer=COMPR_BUFFER)
+        compr_args.update(args.compression)
+        key.compressor = Compressor(**compr_args)
+        decompress_buffer = bytes(COMPR_BUFFER)
+
+        last_segment_id = repository.io.get_latest_segment()
+        segments = list(repository.io.segment_iterator())
+        progress_indicator = ProgressIndicatorPercent(len(segments), start=0, step=0.01, same_line=True,
+                                                      msg="%3.2f %% processed")
+
+        if not repository.io.is_committed_segment(segments[-1][1]):
+            self.print_error('Last segment in repository is uncomitted. Repository corrupted. Try '
+                             '"borg check --repair".')
+            repository.close()
+            return self.exit_code
+        for i, (segment_id, segment_filename) in enumerate(segments):
+            chunks = {}
+            if segment_id <= segment_pointer:
+                # commited segment files are never appended to, yes?
```

Comment:
Thought about it and looked at at the code.

So we have the transaction of the Repository and the one of LoggedIO (the "segment transaction"). So the initial +1 which guarantees that any new transaction will start off in a new segment comes from LoggedIO.cleanup(old_transaction_id), which assigns self.segment = transaction_id + 1.

On Repository.commit() LoggedIO.write_commit() is called - which always closes the current segment. After that compact_segments might squish stuff into less segments than existed before, but because the old segment was closed, a new one is created for these "compacting writes".